ddtrace/contrib/fastapi/patch.py

```diff
@@ -42,6 +42,18 @@ def traced_init(wrapped, instance, args, kwargs):
     wrapped(*args, **kwargs)
 
 
+def trace_add_middleware(wrapped, instance, args, kwargs):
+    # Ensures user_middlewares are added after ddtrace TraceMiddleware
+    if instance.user_middleware and instance.user_middleware[0].cls is TraceMiddleware:
+        # Overrides FastApi.add_middlware(). Note - fastapi.applications.FastAPI is the
+        # child class of starlette.applications.Starlette():
+        # https://github.com/encode/starlette/blob/0.13.2/starlette/applications.py#L115
+        instance.user_middleware.insert(1, Middleware(*args, **kwargs))
+        instance.middleware_stack = instance.build_middleware_stack()
```

Comment:
I worry about this a little bit, since we are basically duplicating the logic from Starlette, if that behavior ever changes then this will break things.

I am not sure a better way at the moment, but calling out that this gives me pause.

Reply:
I think this is worth discussing, as the alternative option is to change the instance's middleware stack after the middleware has been added. Not sure which one is better, as both rely on the internal mechanism of `FastAPI.add_midleware()`.

Reply:
Whenever user middlewares are added [build_middleware_stack](https://github.com/encode/starlette/blob/0.13.2/starlette/applications.py#L58) is called. Instead of patching` Fastapi.__init__()` and `add_middleware()` we can wrap the asgi app returned by `build_middleware_stack` in the ddtrace TraceMiddleware. However this approach patches fastapi "internals" and could break at any point. Also I think it's harder to follow and maintain

Reply:
If I'm following the conversation correctly, there are three options being discussed:

1. This PR's current approach, patching the two-line function `add_middleware` and duplicating its internal logic
2. Wrap the return value of `build_middleware_stack`, a `starlette.routing.Router`
3. Wrap `add_middleware` but use @mabdinur's steps above to avoid duplicating internal logic at the cost of decreased performance

Assuming I understand the options correctly, I'm actually most interested in number 2. This is because my read of the [Starlette documentation suggests](https://www.starlette.io/routing/#working-with-router-instances) that the `starlette.routing.Router` class is part of the public API. Thus, wrapping the instance returned by `build_middleware_stack` appears to be the approach least dependent on the private API.